File: iOS/Native Library Bridging/TSLManager.m
//
//  TSLManager.m
//  chemistry_rfid_mobile_app
//
//  Created by Elliot Francis on 5/30/17.
//  Copyright © 2017 Facebook. All rights reserved.
//

#import "TSLManager.h"
#import <React/RCTLog.h>

@interface TSLManager () {
  NSArray * accessoryList;
  EAAccessory * currentAccessory;
  TSLInventoryCommand * inventoryResponder;
  TSLBarcodeCommand * barcodeResponder;
}

@end

@implementation TSLManager

RCT_EXPORT_MODULE();

// React Function Exports

RCT_REMAP_METHOD(startScannerSearchAsync,
                 startScannerSearchResolver:(RCTPromiseResolveBlock)resolve
                 startScannerSearchRejecter:(RCTPromiseRejectBlock)reject)
{
  [self beginScannerSelect];
  NSMutableArray *results = [NSMutableArray arrayWithCapacity:0];
  for( EAAccessory *accessory in [[EAAccessoryManager sharedAccessoryManager] connectedAccessories]) {
    [results addObject:[NSString stringWithFormat:@"%@", accessory.serialNumber]];
  }
  resolve(results);
}

RCT_EXPORT_METHOD(connectToScannerAtIndex:(NSInteger *)selectedRow)
{
  [self connectToSelectedScanner: (int) selectedRow];
}

RCT_REMAP_METHOD(disconnectScannerAsync,
                 disconnectScannerResolver:(RCTPromiseResolveBlock)resolve
                 disconnectScannerRejecter:(RCTPromiseRejectBlock)reject)
{
  [self haltResponders];
  [self disconnectCommander];
  resolve(@"Scanner Disconnected -- Commander Halted");
}

// Library Support Commands

- (void) initialize {
  [self initCommander];
  [self attachResponders];
}

- (void) initCommander {
  // Initialize Scanner Interface
  self.commander = [[TSLAsciiCommander alloc] init];
  [self.commander addSynchronousResponder];
}

- (void) attachResponders {
  // Attach Responders
  inventoryResponder = [[TSLInventoryCommand alloc] init];
  inventoryResponder.transponderReceivedDelegate = self;
  inventoryResponder.captureNonLibraryResponses = true;
  [self.commander addResponder:inventoryResponder];
  
  barcodeResponder = [[TSLBarcodeCommand alloc] init];
  barcodeResponder.barcodeReceivedDelegate = self;
  barcodeResponder.captureNonLibraryResponses = true;
  [self.commander addResponder:barcodeResponder];
}

- (void) haltResponders {
  [self.commander removeResponder:inventoryResponder];
  [self.commander removeResponder:barcodeResponder];
}

- (void) connectCommander { [self.commander connect:nil]; }

- (void) disconnectCommander { [self.commander halt]; }

- (void) disconnectScanner {
  if(self.commander.isConnected) {
    [self.commander send:@".sl"];
    [self disconnectCommander];
  }
}

- (void) beginScannerSelect {
  
  [[EAAccessoryManager sharedAccessoryManager] registerForLocalNotifications];
  accessoryList = [[EAAccessoryManager sharedAccessoryManager] connectedAccessories];
  [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(_accessoryDidConnect:) name:EAAccessoryDidConnectNotification object:nil];
  [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(_accessoryDidDisconnect:) name:EAAccessoryDidDisconnectNotification object:nil];
}

- (void) connectToSelectedScanner: (int) row {
  
  [self haltResponders];
  [self disconnectCommander];
  [self initCommander];
  [self attachResponders];
  
  currentAccessory = [accessoryList objectAtIndex:row];
  [self.commander connect:currentAccessory];
  if([self.commander isConnected]) {
    [self sendEventWithName:@"tsl-scanner-connected" body:currentAccessory.serialNumber];
  } else {
    [self sendEventWithName:@"tsl-scanner-connect-error" body:@"Unable to Connect to Scanner"];
  }
}


// Library Action Responder Commands

-(void) _accessoryDidConnect:(NSNotification *)notification
{
  EAAccessory *connectedAccessory = [[notification userInfo] objectForKey:EAAccessoryKey];
  if( connectedAccessory.protocolStrings.count != 0 )
  {
    accessoryList = [[EAAccessoryManager sharedAccessoryManager] connectedAccessories];
    NSMutableArray *results = [NSMutableArray arrayWithCapacity:0];
    for( EAAccessory *accessory in [[EAAccessoryManager sharedAccessoryManager] connectedAccessories]) {
      [results addObject:[NSString stringWithFormat:@"%@", accessory.serialNumber]];
    }
    [self sendEventWithName:@"tsl-accessory-list-updated" body:results];
  }
}

- (void)_accessoryDidDisconnect:(NSNotification *)notification
{
  accessoryList = [[EAAccessoryManager sharedAccessoryManager] connectedAccessories];
  NSMutableArray *results = [NSMutableArray arrayWithCapacity:0];
  for( EAAccessory *accessory in [[EAAccessoryManager sharedAccessoryManager] connectedAccessories]) {
    [results addObject:[NSString stringWithFormat:@"%@", accessory.serialNumber]];
  }
  [self sendEventWithName:@"tsl-accessory-list-updated" body:results];
}

- (void) transponderReceived:(NSString *)epc crc:(NSNumber *)crc pc:(NSNumber *)pc rssi:(NSNumber *)rssi fastId:(NSData *)fastId moreAvailable:(BOOL)moreAvailable {
    [self sendEventWithName:@"tsl-transponder-received" body:epc];
}

- (void) barcodeReceived:(NSString *)data {
    [self sendEventWithName:@"tsl-barcode-received" body:data];
}


- (NSArray *) supportedEvents {
  return @[
           @"tsl-transponder-received",
           @"tsl-barcode-received",
           @"tsl-accessory-list-updated",
           @"tsl-scanner-connected",
           @"tsl-scanner-disconnected",
           @"tsl-scanner-connect-error"];
}
@end

File: iOS/TSLManager.swift
//
//  TSLManager.swift
//  Bridging Module for interfacing a Swift Based iOS Application with
//  the TSL 1128 Barcode Scanner.
//
//  Created by Elliot Francis on 4/18/16.
//  Copyright © 2016 University of Utah. All rights reserved.
//

protocol TSLScannerResponderDelegate {
    func didRecieveBarcode(_ barcode:String);
    func didRecieveTransponder(_ epc:String, moreAvailable:Bool);
}

protocol TSLScannerSelectionDelegate {
    func accessoryListDidUpdate(_ accessoryList: NSArray);
    func didSelectScanner(_ name: String);
}

protocol TSLScannerStatusDelegate {
    func scannerDidDisconnect()
}

import Foundation
class TSLManager: NSObject, TSLInventoryCommandTransponderReceivedDelegate, TSLBarcodeCommandBarcodeReceivedDelegate {

    static var sharedInstance = TSLManager();

    // Manager Variables
    var _commander:TSLAsciiCommander?
    var _currentAccessory:EAAccessory?
    var _accessoryList = [EAAccessory]();
    var _inventoryResponder: TSLInventoryCommand?;
    var _barcodeResponder: TSLBarcodeCommand?;

    // Action Delegates
    var scannerDelegate: TSLScannerResponderDelegate?;
    var selectionDelegate: TSLScannerSelectionDelegate?;
    var statusDelegate: TSLScannerStatusDelegate?;

    // MARK: - Initialization Function
    func initialize() {
        self.initCommander();
        self.attachResponders();
    }

    // MARK: - Scanner Management Support Functions
    func initCommander() -> Void {
        _commander = TSLAsciiCommander();
        //_commander?.addResponder(TSLLoggerResponder());
        _commander?.addSynchronousResponder();
    }

    func attachResponders() {
        _inventoryResponder = TSLInventoryCommand()
        _inventoryResponder!.transponderReceivedDelegate = self;
        _inventoryResponder!.captureNonLibraryResponses = true;
        _commander!.add(_inventoryResponder);

        _barcodeResponder = TSLBarcodeCommand()
        _barcodeResponder!.barcodeReceivedDelegate = self;
        _barcodeResponder!.captureNonLibraryResponses = true;
        _commander!.add(_barcodeResponder);
    }

    func haltResponders() {
        _commander!.remove(_inventoryResponder);
        _commander!.remove(_barcodeResponder);
    }

    func haltCommander() {
        _commander!.halt();
    }

    func connectCommander() {
        _commander?.connect(nil);
    }

    func disconnectCommander() {
        _commander?.disconnect()
    }

    func disconnectScanner() {
        if((_commander?.isConnected)!) {
            _commander?.send(".sl");
            self.disconnectCommander();
        }
    }

    func terminateCommander() {
        self.haltResponders();
        self.haltCommander();
        _commander = nil;

    }

    func setScannerPower(_ powerLevel: Int32) {
        if( self._commander!.isConnected )
        {
            let command:TSLInventoryCommand = TSLInventoryCommand.synchronousCommand();
            command.takeNoAction = TSL_TriState_YES;
            command.outputPower = powerLevel;
            self._commander!.execute(command);
        }
    }

    // MARK: - Device Selection Support Functions
    func beginScannerSelect() {
        EAAccessoryManager.shared().registerForLocalNotifications();
        _accessoryList = EAAccessoryManager.shared().connectedAccessories;
        print("List Accessories")
        print(_accessoryList)

        NotificationCenter.default.addObserver(self, selector: #selector(self.accessoryDidConnect), name: NSNotification.Name.EAAccessoryDidConnect, object: nil);
        NotificationCenter.default.addObserver(self, selector: #selector(self.accessoryDidDisconnect), name: NSNotification.Name.EAAccessoryDidDisconnect, object: nil);
    }

    func accessoryDidConnect(_ notification: Notification) {
        let userInfo: NSDictionary? = (notification as NSNotification).userInfo as NSDictionary?
        let connectedAccessory = userInfo!.object(forKey: EAAccessoryKey)
        if ( (connectedAccessory as AnyObject).protocolStrings.count != 0) {
            _accessoryList = EAAccessoryManager.shared().connectedAccessories
            self.selectionDelegate?.accessoryListDidUpdate(_accessoryList as NSArray);
        }
    }

    func selectScannerForRow(_ index: Int) {
        self.disconnectCommander();
        _currentAccessory = _accessoryList[_accessoryList.startIndex.advanced(by: index)];
        let didConnect = (_commander?.connect(_currentAccessory))! as Bool;
        if(didConnect) {
            self.selectionDelegate!.didSelectScanner((_currentAccessory?.serialNumber)!);
        }
    }

    func refreshList() {
        _accessoryList = EAAccessoryManager.shared().connectedAccessories
        self.selectionDelegate?.accessoryListDidUpdate(_accessoryList as NSArray);
    }


    func accessoryDidDisconnect(_ notification: Notification) {
        _accessoryList = EAAccessoryManager.shared().connectedAccessories;
    }

    // MARK: - TSL Responder Delegates
    func transponderReceived(_ epc: String!, crc: NSNumber!, pc: NSNumber!, rssi: NSNumber!, fastId: Data!, moreAvailable: Bool) {
        //print("Debug 1");
        //print("epc : \(epc)");
        //print("rssi : \(rssi)");
        //print("moreAvailable : \(moreAvailable)");

        self.scannerDelegate?.didRecieveTransponder(epc, moreAvailable: moreAvailable);
    }

    func barcodeReceived(_ data: String!) {
        self.scannerDelegate?.didRecieveBarcode(data);
    }


}
